Mathlib/Analysis/Normed/Field/Ultra.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.Normed.Group.Ultra
+
+/-!
+## Sufficient contidition to have an ultrametric norm on a field
+
+This file provides ways of constructing an instance of `IsUltrametricDist` based on
+facts about the existing norm.
+
+## Main results
+
+* `isUltrametricDist_of_forall_norm_natCast_le_one`: a norm in a field is ultrametric if the norm
+  of the image of a natural is less than or equal to one
+
+## Implementation details
+
+The proof relies on a bounded-from-above argument, so the imports need to include `rpow`
+
+## Tags
+
+ultrametric, nonarchimedean
+-/
+open Metric IsUltrametricDist NNReal
+
+namespace IsUltrametricDist
+
+section sufficient
+
+variable {R F : Type*} [NormedDivisionRing R] [NormedField F]
+
+lemma isUltrametricDist_of_forall_norm_add_one_le_max_norm_one
+    (h : ∀ x : R, ‖x + 1‖ ≤ max ‖x‖ 1) : IsUltrametricDist R := by
+  apply isUltrametricDist_of_forall_norm_add_le_max_norm
+  intro x y
+  rcases eq_or_ne y 0 with rfl|hy
+  · simp
+  rw [← div_le_div_right (c := ‖y‖) (by simpa using hy), ← norm_div, add_div, div_self hy,
+      ← max_div_div_right (norm_nonneg _), div_self (by simp [hy]), ← norm_div]
+  exact h _
+
+lemma isUltrametricDist_of_forall_norm_sub_one_of_norm_le_one
+    (h : ∀ x : R, ‖x‖ ≤ 1 → ‖x - 1‖ ≤ 1) : IsUltrametricDist R := by
+  have : ∀ x : R, ‖x‖ ≤ 1 → ‖x + 1‖ ≤ 1 := by
+    intro x hx
+    specialize h (-x) (by simpa using hx)
+    rwa [← neg_add', norm_neg] at h
+  apply isUltrametricDist_of_forall_norm_add_one_le_max_norm_one
+  intro x
+  cases le_or_lt ‖x‖ 1 with
+  | inl h => simpa [h] using this _ h
+  | inr h =>
+    suffices ‖x + 1‖ ≤ ‖x‖ by simp [this]
+    rw [← div_le_div_right (c := ‖x‖) (h.trans' zero_lt_one), div_self (h.trans' zero_lt_one).ne',
+        ← norm_div, add_div, div_self (by simpa using (h.trans' zero_lt_one)), add_comm]
+    apply this
+    simp [inv_le_one_iff₀, h.le]
+
+/-- To prove that a normed field is nonarchimedean, it suffices to prove that the norm
+of the image of any natural is less than or equal to one. -/
+lemma isUltrametricDist_of_forall_norm_natCast_le_one
+    (h : ∀ n : ℕ, ‖(n : F)‖ ≤ 1) : IsUltrametricDist F := by
+  -- we first use our hypothesis about the norm of naturals to have that multiplication by
+  -- naturals keeps the norm small
+  replace h : ∀ (x : F) (n : ℕ), ‖n • x‖ ≤ ‖x‖ := by
```

Comment:
You can always golf `have h : ∀ (x : foo), bar := by intro x; ...` to `have h (x : foo) : bar`, and the same with `replace`. (Doesn't work with `suffices` though, unfortunately.)